### PR TITLE
improve: allow quotes for non-messages if relayer balance is too low

### PR DESCRIPTION
**Pull request by james-a-morris**
We received a ticket that a large token holder couldn't receive a quote due to the relayer not having enough balance. This is the case when we need to partially fill the deposit & the simulated gas cost does not take this into account today. 

To circumvent this issue, we can pass a balance amount to be used for simulating the relay. This can be set optionally to be a full fill (in the event that the caller knows there is a message). By default, this is set to 1wei to simulate a zeroFill.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@across-protocol/sdk-v2",
   "author": "UMA Team",
-  "version": "0.17.8",
+  "version": "0.17.9",
```
Comment (pxrl) [inline package.json:4]:
```suggestion
  "version": "0.17.10",
```

**Review by pxrl — APPROVED** [review]

File: src/relayFeeCalculator/relayFeeCalculator.ts [hunk 2]
```diff
@@ -197,6 +198,7 @@ export class RelayFeeCalculator {
   async gasFeePercent(
     deposit: Deposit,
     amountToRelay: BigNumberish,
+    amountToSimulate: BigNumberish,
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts]: Updated to make this a boolean that defaults to true. I agree that we may want to add an override type to these functions, but it may be better suited in a separate PR.

File: src/relayFeeCalculator/relayFeeCalculator.ts [hunk 3]
```diff
@@ -284,6 +286,7 @@ export class RelayFeeCalculator {
   async relayerFeeDetails(
     deposit: Deposit,
     amountToRelay?: BigNumberish,
+    amountToSimulate = bnOne as BigNumberish,
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts]: Removed in lieu of a simple approcah

File: src/relayFeeCalculator/relayFeeCalculator.ts [hunk 4]
```diff
@@ -207,10 +228,12 @@ export class RelayFeeCalculator {
...
     if (toBN(amountToRelay).eq(0)) return MAX_BIG_INT;
 
-    const getGasCosts = this.queries.getGasCosts(deposit, amountToRelay, relayerAddress).catch((error) => {
-      this.logger.error({ at: "sdk-v2/gasFeePercent", message: "Error while fetching gas costs", error });
-      throw error;
-    });
+    const getGasCosts = this.queries
+      .getGasCosts(deposit, simulateZeroFill ? bnOne : amountToRelay, relayerAddress)
+      .catch((error) => {
+        this.logger.error({ at: "sdk-v2/gasFeePercent", message: "Error while fetching gas costs", error });
```
Comment (pxrl) [inline src/relayFeeCalculator/relayFeeCalculator.ts:234]: While here, should we include in this log message?
 - deposit information
 - simulateZeroFill flag

**Review by pxrl — APPROVED** [review]
This looks pretty clean :+1:

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 4 above)
Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts]: Done: https://github.com/across-protocol/sdk-v2/pull/445/commits/901d590328278ef5e1fb2e7c4aa03e4614ea0823

File: src/relayFeeCalculator/relayFeeCalculator.ts [hunk 5]
```diff
@@ -207,12 +229,26 @@ export class RelayFeeCalculator {
...
+        throw error;
+      });
     const getTokenPrice = this.queries.getTokenPrice(tokenInformation.symbol).catch((error) => {
-      this.logger.error({ at: "sdk-v2/gasFeePercent", message: "Error while fetching token price", error });
+      this.logger.error({
+        at: "sdk-v2/gasFeePercent",
+        message: "Error while fetching token price",
+        error,
+        simulateZeroFill,
+        deposit: JSON.stringify(deposit, jsonReplacerWithBigNumbers),
```
Comment (pxrl) [inline src/relayFeeCalculator/relayFeeCalculator.ts:250]: `simulateZeroFill` and the complete `deposit` object aren't really necessary here - how about just logging the `destinationToken` ?

File: src/relayFeeCalculator/relayFeeCalculator.ts [hunk 6]
```diff
@@ -194,9 +195,29 @@ export class RelayFeeCalculator {
...
+   *       on a zero fill. However, the percentage will be returned as a percentage of the
+   *       amount to relay. This is useful for determining the maximum gas fee % that a
+   *       relayer may need to make on a regular fill. You will get differing results if
+   *       a message & recipient contract is provided as this function may not simulate with
+   *       the correct parameters to see a full fill.
+   */
   async gasFeePercent(
     deposit: Deposit,
     amountToRelay: BigNumberish,
+    simulateZeroFill = true,
```
Comment (nicholaspai) [inline src/relayFeeCalculator/relayFeeCalculator.ts:220]: As mentioned offline, we should rename this to indicate its a 1 wei fill and also simulate a full fill instead of a partial

Comment (pxrl) [reply src/relayFeeCalculator/relayFeeCalculator.ts:220]: fwiw, just for alignment, "zero fill" is the term used in the relayer to indicate a 1 wei fill (counter-intuitive). As I understand it, zero-fill also means 1 wei in this context.

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 5 above)
Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts]: Done

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 6 above)
Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts:220]: @nicholaspai this is using the `relayer-v2` terminology of a `zeroFill` being `1wei`. Setting `simulateZeroFill = false` will attempt to `fillRelay( )` with `amountToRelay` as the `MaxTokensToSend`.

Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts:220]: Should we add an additional flag to override this so that `amountToRelay` is ignored for the simulation purpose?

Comment (nicholaspai) [reply src/relayFeeCalculator/relayFeeCalculator.ts:220]: oki'm on board with naming it `zeroFill` and no I don't think another flag is the correct way to do this. 

Why not do:

- if `simulateZeroFill` is `true`:
- pass in `{ ...deposit, amount: 1 wei }` when simulating the gas costs to simulate the full fill and then you can still derive `gasFeePercent` from `gasCost / amountToRelay`?

Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts:220]: I think I'd be okay with this. 

As an alternative, per @mrice32 we can usually assume that if a message is non-empty, then we should attempt to make a full fill, but if the message is not empty, we can default back to the original usage and attempt to fill with one unit. 

In the above case, we could also fix deposit.amount = 1. 

It's a shame that we moved the check against partial fills outside of the SDK because that would be a good decider of whether or not we should reject if the relayer balance is not enough to full fill a message.

TLDR: THoughts on removing `simulateZeroFill` in lieu of checking if `isMessageEmpty(deposit)` as the conditional for whether or not we attempt to simulate with a zero fill?

Comment (mrice32) [reply src/relayFeeCalculator/relayFeeCalculator.ts:220]: I like having an explicit flag, personally. I don't love the idea of inferring based on other variables. We should be consistent on deciding this outside of this function or inside, and we've decided that outside is the way to go.

Secondly, 1 wei fills will probably not give an accurate estimation. They can round to 0, causing the math to change and no tokens to be transferred. I would suggest 100 units (small no matter which token) or similar. There _may not be_ rounding if the fees are all 0 with 1 wei, but I'm not 100% sure (we could test this). Thoughts on either testing this or making it 100 units instead of 1?

**Review by mrice32 — COMMENTED** [review]
Looks good, just one comment!

**Comment by james-a-morris** [comment]
@nicholaspai @mrice32 I've updated this PR so that our zero fills are making full fills rather than partial. This differs from the terminology used in `relayer-v2` in that `zeroFills` can initiate a slow fill by relaying `1wei`.

File: src/relayFeeCalculator/relayFeeCalculator.ts [hunk 7]
```diff
@@ -207,12 +227,33 @@ export class RelayFeeCalculator {
 
     if (toBN(amountToRelay).eq(0)) return MAX_BIG_INT;
 
-    const getGasCosts = this.queries.getGasCosts(deposit, amountToRelay, relayerAddress).catch((error) => {
-      this.logger.error({ at: "sdk-v2/gasFeePercent", message: "Error while fetching gas costs", error });
-      throw error;
-    });
+    const getGasCosts = this.queries
+      .getGasCosts(
```
Comment (mrice32) [inline src/relayFeeCalculator/relayFeeCalculator.ts:231]: Sorry, one last question: did you change this simulation to using nonzero fees (1 wei should be fine), just to match the previous behavior?

**Review by mrice32 — COMMENTED** [review]
Just one more comment!

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 7 above)
Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts:231]: It matches the deposit, but the API does pass in a non-zero fee. It's visible in the PR on the FE.

File: src/relayFeeCalculator/relayFeeCalculator.ts [hunk 8]
```diff
@@ -194,9 +194,29 @@ export class RelayFeeCalculator {
...
+   *       on a zero fill. However, the percentage will be returned as a percentage of the
+   *       amount to relay. This is useful for determining the maximum gas fee % that a
+   *       relayer may need to make on a regular fill. You will get differing results if
+   *       a message & recipient contract is provided as this function may not simulate with
+   *       the correct parameters to see a full fill.
+   */
   async gasFeePercent(
     deposit: Deposit,
     amountToRelay: BigNumberish,
+    simulateZeroFill = true,
```
Comment (nicholaspai) [inline src/relayFeeCalculator/relayFeeCalculator.ts:219]: why make this default `true`? I would think this should be default `false`

**Review by nicholaspai — COMMENTED** [review]
One question

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 7 above)
Comment (mrice32) [reply src/relayFeeCalculator/relayFeeCalculator.ts:231]: Gotcha, okay, cool.

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 8 above)
Comment (mrice32) [reply src/relayFeeCalculator/relayFeeCalculator.ts:219]: Good point -- agreed.

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 6 above)
Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts:219]: Resolved offline with @mrice32

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 8 above)
Comment (james-a-morris) [reply src/relayFeeCalculator/relayFeeCalculator.ts]: My thought is that our default case so far is that we're dealing with non-message fields. However, I am of two minds.

I've updated the PR to make this default to false https://github.com/across-protocol/sdk-v2/pull/445/commits/801cfaac30169e380f6aa5b65191a23cbe423bb4

**Review by mrice32 — APPROVED** [review]
LGTM!

File: src/relayFeeCalculator/relayFeeCalculator.ts (hunk 8 above)
Comment (mrice32) [reply src/relayFeeCalculator/relayFeeCalculator.ts]: That may be the common case for the API, but other users, like the relayer, would probably want to do this estimate on the un-modified deposit, right? I think the setting that is what an uninformed caller might expect is probably the better default.
